pkg/cable/wireguard/driver.go

```diff
@@ -0,0 +1,411 @@
+	}
+	pub = priv.PublicKey()
+	if localEndpoint.Spec.BackendConfig == nil {
+		localEndpoint.Spec.BackendConfig = make(map[string]string)
+	}
+	localEndpoint.Spec.BackendConfig[PublicKey] = pub.String()
+
+	// configure the device. still not up
+	port := DefaultListenPort
+	peerConfigs := make([]wgtypes.PeerConfig, 0)
+	cfg := wgtypes.Config{
+		PrivateKey:   &priv,
+		ListenPort:   &port,
+		FirewallMark: nil,
+		ReplacePeers: true,
+		Peers:        peerConfigs,
+	}
+	if err = wg.client.ConfigureDevice(DefaultDeviceName, cfg); err != nil {
+		return nil, fmt.Errorf("failed to configure WireGuard device: %v", err)
+	}
+
+	klog.V(log.DEBUG).Infof("Initialized WireGuard %s with publicKey %s", DefaultDeviceName, pub)
+	return &wg, nil
+}
+
+func (w *wireguard) Init() error {
+	// ip link set $DefaultDeviceName up
+	if err := netlink.LinkSetUp(w.link); err != nil {
+		return fmt.Errorf("failed to bring up WireGuard device: %v", err)
+	}
+	return nil
+}
+
+func (w *wireguard) ConnectToEndpoint(remoteEndpoint types.SubmarinerEndpoint) (string, error) {
+
+	if w.localEndpoint.Spec.ClusterID == remoteEndpoint.Spec.ClusterID {
+		klog.V(log.DEBUG).Infof("Will not connect to self")
+		return "", nil
+	}
+
+	// parse remote addresses and allowed IPs
+	ip := endpointIP(&remoteEndpoint)
+	remoteIP := net.ParseIP(ip)
+	if remoteIP == nil {
+		return "", fmt.Errorf("failed to parse remote IP %s", ip)
+	}
+	allowedIPs := parseSubnets(remoteEndpoint.Spec.Subnets)
+
+	// parse remote public key
+	remoteKey, err := publicKey(remoteEndpoint)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse peer public key: %v", err)
+	}
+
+	klog.V(log.DEBUG).Infof("Connecting cluster %s endpoint %s with publicKey %s", remoteEndpoint.Spec.ClusterID, remoteIP, remoteKey)
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	// delete old peers for ClusterID
+	oldKey, found := w.peers[remoteEndpoint.Spec.ClusterID]
+	if found {
+		if oldKey.String() == remoteKey.String() {
+			//TODO check that peer config has not changed (eg allowedIPs)
+			klog.V(log.DEBUG).Infof("Skipping update of existing peer key %s", oldKey)
+			return ip, nil
+		}
+
+		// new peer will take over subnets so can ignore error
+		_ = removePeer(w.client, &oldKey)
+
+		delete(w.peers, remoteEndpoint.Spec.ClusterID)
+	}
+
+	// configure peer
+	peerCfg := []wgtypes.PeerConfig{{
+		PublicKey:    *remoteKey,
+		Remove:       false,
+		UpdateOnly:   false,
+		PresharedKey: nil,
+		Endpoint: &net.UDPAddr{
+			IP:   remoteIP,
+			Port: DefaultListenPort,
+		},
+		PersistentKeepaliveInterval: nil,
+		ReplaceAllowedIPs:           true,
+		AllowedIPs:                  allowedIPs,
+	}}
+	err = w.client.ConfigureDevice(DefaultDeviceName, wgtypes.Config{
+		ReplacePeers: false,
+		Peers:        peerCfg,
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to configure peer: %v", err)
+	}
+
+	// verify peer was added
+	// TODO verify configuration
+	if p, err := peerByKey(w.client, remoteKey); err != nil {
```

Comment:
Shouldn't success from `ConfigureDevice` above be enough to know that the peer was added?

Reply:
It should, but we had some silent failures in our tests. So added this for better debugging.
Peer configurations is not that common, so I don't think it is a problem to verify.